File: 2020/Excercise 4 - Valid Passports/golang/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strings"
	"io"
)

type Passport struct {
	BirthYear string
	IssueYear string
	ExpirationYear string
	Height string
	HairColor string
	EyeColor string
	PassportId string
	CountryId string
}

func (p Passport) IsValidPassport() (bool) {
	return p.IsAllFieldsRequired() || p.Is7FieldsRequired()
}
func (p Passport) IsAllFieldsRequired() (bool) {
	return (p.BirthYear != "" &&  p.IssueYear != "" && p.ExpirationYear != "" && p.Height != "" && p.EyeColor != "" && p.PassportId != "" && p.CountryId != "" && p.HairColor != "")
}

func (p Passport) Is7FieldsRequired() (bool) {
	return (p.BirthYear != "" &&  p.IssueYear != "" && p.ExpirationYear != "" && p.Height != "" && p.EyeColor != "" && p.PassportId != ""  && p.HairColor != "")
}

func LoadPassports(path string) ([]Passport) {
	
	lines := readLines(path)
	
	var passports []Passport
	var sb strings.Builder

	for _, l := range lines {
		text := strings.Trim(l, "\n\r")
		if text != "" {	
			sb.WriteString(text)
			sb.WriteString(" ")
		} else {
			var str = sb.String()
			strArray := strings.Split(str," ")

			
			p := Passport {
				BirthYear: Find(strArray, "byr"),
				IssueYear: Find(strArray, "iyr"),
				ExpirationYear: Find(strArray, "eyr"),
				Height: Find(strArray, "hgt"),
				HairColor: Find(strArray, "hcl"),
				EyeColor: Find(strArray, "ecl"),
				PassportId: Find(strArray, "pid"),
				CountryId: Find(strArray, "cid"),
			}
			passports = append(passports, p)
			sb.Reset()
		}
	}

	if sb.String() != "" {
		var str = sb.String()
			strArray := strings.Split(str," ")

			
			p := Passport {
				BirthYear: Find(strArray, "byr"),
				IssueYear: Find(strArray, "iyr"),
				ExpirationYear: Find(strArray, "eyr"),
				Height: Find(strArray, "hgt"),
				HairColor: Find(strArray, "hcl"),
				EyeColor: Find(strArray, "ecl"),
				PassportId: Find(strArray, "pid"),
				CountryId: Find(strArray, "cid"),
			}
			passports = append(passports, p)
			sb.Reset()
	}

	return passports
}


// Find returns the smallest index i at which x == a[i],
// or len(a) if there is no such index.
func Find(a []string, x string) string {
	for _, n := range a {
		if strings.Contains(n, x) {
			strA := strings.Split(n, ":")
			return strA[1]
		}
	}
	
	return ""
}


func readLines(path string) (lines []string) {
	file, err := os.Open(path)
	if err != nil {
		panic(fmt.Sprintf("Cannot open file: %v", err))
	}

	defer file.Close()

	reader := bufio.NewReader(file)
	
	for {
		line, err := reader.ReadString('\n')
		if err != nil && err != io.EOF {
			panic(fmt.Sprintf("Cannot read line: %v", err))
		}

			
		lines = append(lines, line)

		if err == io.EOF {
			break
		}
	}
	
	return lines

}

func main() {
	os.Chdir("../input")
	curPath, err := os.Getwd()
	
	if err != nil {
		log.Fatal("main: %s", err)
	}

	passports := LoadPassports(curPath + "\\day4input.dat")

	var totalValidPassports int
	for _, p := range passports {
		if (p.IsValidPassport()) {
			totalValidPassports++
		}
	}

	fmt.Println("Total Valid Passports %d", totalValidPassports)

	
}

File: 2020/Excercise 4 - Valid Passports/python/main.py
#!/usr/bin/env python

import re

with open('./input', 'r') as input:
    passport_data = [l.strip() for l in input.readlines()]

def parse_passports(passports):
    passport = []
    for line in passports + ['']:
        if line == '' and len(passport) > 0:
            yield ' '.join(passport).strip()
            passport = []
        passport.append(line)

def part1():
    passports = parse_passports(passport_data)
    valid_count = 0
    for passport in passports:
        parts = passport.split()
        parts_count = len(parts)
        if ((parts_count == 8) or
           (parts_count == 7 and all(map(lambda s: not s.startswith('cid'), parts)))):
            valid_count = valid_count + 1
    print(f'Number of valid passports = {valid_count}')

def part2():
    valid_byr = lambda byr: (byr != None and int(byr) in range(1920, 2002+1)) 
    valid_iyr = lambda iyr: (iyr != None and int(iyr) in range(2010, 2020+1))
    valid_eyr = lambda eyr: (eyr != None and int(eyr) in range(2020, 2030+1))
    valid_hgt = lambda hgt: (hgt != None and re.match('^(1([5-8][0-9]|9[0-3])cm|(59|6[0-9]|7[0-6])in)$', hgt))
    valid_hcl = lambda hcl: (hcl != None and re.match('^#[0-9a-f]{6}$', hcl))
    valid_ecl = lambda ecl: (ecl != None and ecl in ['amb', 'blu', 'brn', 'gry', 'grn', 'hzl', 'oth'])
    valid_pid = lambda pid: (pid != None and re.match('^[0-9]{9}$', pid))

    def validate_parts(byr=None, iyr=None, eyr=None, hgt=None, hcl=None, ecl=None, pid=None, cid=None):
        return (
            valid_byr(byr) and
            valid_iyr(iyr) and
            valid_eyr(eyr) and
            valid_hgt(hgt) and
            valid_hcl(hcl) and
            valid_ecl(ecl) and
            valid_pid(pid)
        )
    passports = parse_passports(passport_data)
    valid_count = 0
    for passport in passports:
        fields = {x[0]: x[1] for field in passport.split() for x in [field.split(':')]}
        if validate_parts(**fields):
            valid_count = valid_count + 1
    print(f'Number of valid passports = {valid_count}')

part1()
part2()